Multiple versions

Martinsos opened the pull request:
Fixes https://github.com/wasp-lang/wasp/issues/796 .

TODO: Update documentation!

shayneczyzewski commented:
Wow, this is so cool @Martinsos! Great work! I think it will be so amazing when we can tell people to install a specific version, like: `curl -sSL https://raw.githubusercontent.com/wasp-lang/get-wasp/multiple-versions/installer.sh | sh -s -- -v 0.5.0.0` I am no shell script expert, so I just have a few comments/questions:
- I ran the code through a linter (https://www.shellcheck.net/) and it gave back the following. Not sure if any are worth fixing, but just FYI:
```bash
$ shellcheck myscript
 
[Line 70:](javascript:setPosition(70, 20))
    LATEST_VERSION=`get_latest_wasp_version`
                   ^-- [SC2006](https://www.shellcheck.net/wiki/SC2006) (style): Use $(...) notation instead of legacy backticks `...`.

Did you mean: ([apply this](javascript:applyFixIndex([0])), apply [all SC2006](javascript:applyFixCode(2006)))
    LATEST_VERSION=$(get_latest_wasp_version)
 
[Line 134:](javascript:setPosition(134, 29))
        info "      ${BOLD}"'export PATH=$PATH:'"$BIN_DST_DIR${RESET}"
                            ^-- [SC2016](https://www.shellcheck.net/wiki/SC2016) (info): Expressions don't expand in single quotes, use double quotes for that.
 
[Line 170:](javascript:setPosition(170, 12))
    printf "${RED}$@${RESET}\n" >&2
           ^-- [SC2059](https://www.shellcheck.net/wiki/SC2059) (info): Don't use variables in the printf format string. Use printf '..%s..' "$foo".
                  ^-- [SC2145](https://www.shellcheck.net/wiki/SC2145) (error): Argument mixes string and array. Use * or separate argument.
 
[Line 175:](javascript:setPosition(175, 12))
    printf "$@\n"
           ^-- [SC2059](https://www.shellcheck.net/wiki/SC2059) (info): Don't use variables in the printf format string. Use printf '..%s..' "$foo".
            ^-- [SC2145](https://www.shellcheck.net/wiki/SC2145) (error): Argument mixes string and array. Use * or separate argument.
 
[Line 218:](javascript:setPosition(218, 5))
    local PATH_NORMALIZED=$(printf '%s' "$PATH" | sed -e "s|:~|:$HOME|g" | sed -e "s|^~|$HOME|")
    ^-- [SC3043](https://www.shellcheck.net/wiki/SC3043) (warning): In POSIX sh, 'local' is undefined.
          ^-- [SC2155](https://www.shellcheck.net/wiki/SC2155) (warning): Declare and assign separately to avoid masking return values.
                          ^-- [SC2046](https://www.shellcheck.net/wiki/SC2046) (warning): Quote this to prevent word splitting.
 
[Line 221:](javascript:setPosition(221, 5))
    local QUERY_NORMALIZED=$(printf '%s' "$1" | sed -e "s|^~|$HOME|")
    ^-- [SC3043](https://www.shellcheck.net/wiki/SC3043) (warning): In POSIX sh, 'local' is undefined.
          ^-- [SC2155](https://www.shellcheck.net/wiki/SC2155) (warning): Declare and assign separately to avoid masking return values.
                           ^-- [SC2046](https://www.shellcheck.net/wiki/SC2046) (warning): Quote this to prevent word splitting.
 
[Line 228:](javascript:setPosition(228, 125))
    DATA='{ "api_key": "'$POSTHOG_WASP_PUBLIC_API_KEY'", "type": "capture", "event": "install-script:run", "distinct_id": "'$RANDOM`date +'%s%N'`'", "properties": { "os": "'`get_os_info`'" } }'
>>                                                                                                                          ^-- [SC3028](https://www.shellcheck.net/wiki/SC3028) (warning): In POSIX sh, RANDOM is undefined.
>>                                                                                                                                 ^-- [SC2006](https://www.shellcheck.net/wiki/SC2006) (style): Use $(...) notation instead of legacy backticks `...`.
>>                                                                                                                                                                           ^-- [SC2006](https://www.shellcheck.net/wiki/SC2006) (style): Use $(...) notation instead of legacy backticks `...`.

Did you mean: ([apply this](javascript:applyFixIndex([13,14])), apply [all SC2006](javascript:applyFixCode(2006)))
    DATA='{ "api_key": "'$POSTHOG_WASP_PUBLIC_API_KEY'", "type": "capture", "event": "install-script:run", "distinct_id": "'$RANDOM$(date +'%s%N')'", "properties": { "os": "'$(get_os_info)'" } }'
 
[Line 243:](javascript:setPosition(243, 36))
        curl -LIs -o /dev/null -w %{url_effective} https://github.com/wasp-lang/wasp/releases/latest | awk -F/ '{print $NF}' | cut -c2-
                                   ^-- [SC1083](https://www.shellcheck.net/wiki/SC1083) (warning): This { is literal. Check expression (missing ;/\n?) or quote it.
                                                 ^-- [SC1083](https://www.shellcheck.net/wiki/SC1083) (warning): This } is literal. Check expression (missing ;/\n?) or quote it.

$
```
- Did we always install the data files, in addition to the executable, before? What do we need those for again? I thought that was packaged into the binary.
- Do we want to have a nicer error message when they try to install a version that cannot be found?
<img width="862" alt="Screenshot 2022-11-01 104823" src="https://user-images.githubusercontent.com/523636/199262169-3d169014-fb31-442e-8bde-a8f446d9dbeb.png">

Overall, awesome job! I also left a couple of small ideas on tweaking the language, but looking great! 🚀

shayneczyzewski commented on installer.sh:
```diff
@@ -63,63 +63,66 @@ get_os_info() {
+        LATEST_VERSION_MESSAGE="latest is $LATEST_VERSION"
+    fi
+
+    info "Installing wasp v$VERSION_TO_INSTALL ($LATEST_VERSION_MESSAGE).\n"
```
Do we want to standardize on using a v prefix vs not in user-level messaging?

`Installing wasp v0.5.1.0 (latest is 0.6.0.0).`

In other places we say "version", so maybe here we say "Installing wasp version XYZ"? Super small, just noting.

Martinsos replied:
Done, now using `version`!

shayneczyzewski commented on installer.sh:
```diff
@@ -63,63 +63,66 @@ get_os_info() {
-    if [ ! -z "$OCCUPIED_PATH_ERRORS" ]; then
-        die "\nInstallation failed!\n\n${OCCUPIED_PATH_ERRORS}\nRemove listed entries manually or run the installer with --force flag to write over them:\n\n  ${BOLD}curl -sSL https://get.wasp-lang.dev/installer.sh | sh -s -- --force${RESET}\n"
+    else
+        info "Found already existing installation on the disk, at $DATA_DST_DIR. Using it instead.\n"
```
```suggestion
        info "Found an existing installation on the disk, at $DATA_DST_DIR. Using it instead.\n"
```

Martinsos replied:
Done!

shayneczyzewski commented on installer.sh:
```diff
@@ -63,63 +63,66 @@ get_os_info() {
+    if [ -e "$BIN_DST_DIR/wasp" ]; then
+        info "Configuring Wasp executable at $BIN_DST_DIR/wasp to use wasp version $VERSION_TO_INSTALL."
+    else
+        info "Installing Wasp executable to $BIN_DST_DIR/wasp."
```
Super small, but do we want to use "Wasp" or "wasp" throughout? It is mixed currently.

Martinsos replied:
Done, now it is all `wasp`!

Martinsos commented:
> Wow, this is so cool @Martinsos! Great work! I think it will be so amazing when we can tell people to install a specific version, like: `curl -sSL https://raw.githubusercontent.com/wasp-lang/get-wasp/multiple-versions/installer.sh | sh -s -- -v 0.5.0.0` I am no shell script expert, so I just have a few comments/questions:
> 
> * I ran the code through a linter (https://www.shellcheck.net/) and it gave back the following. Not sure if any are worth fixing, but just FYI:
> 
> ```shell
> $ shellcheck myscript
>  
> [Line 70:](javascript:setPosition(70, 20))
>     LATEST_VERSION=`get_latest_wasp_version`
>                    ^-- [SC2006](https://www.shellcheck.net/wiki/SC2006) (style): Use $(...) notation instead of legacy backticks `...`.
> 
> Did you mean: ([apply this](javascript:applyFixIndex([0])), apply [all SC2006](javascript:applyFixCode(2006)))
>     LATEST_VERSION=$(get_latest_wasp_version)
>  
> [Line 134:](javascript:setPosition(134, 29))
>         info "      ${BOLD}"'export PATH=$PATH:'"$BIN_DST_DIR${RESET}"
>                             ^-- [SC2016](https://www.shellcheck.net/wiki/SC2016) (info): Expressions don't expand in single quotes, use double quotes for that.
>  
> [Line 170:](javascript:setPosition(170, 12))
>     printf "${RED}$@${RESET}\n" >&2
>            ^-- [SC2059](https://www.shellcheck.net/wiki/SC2059) (info): Don't use variables in the printf format string. Use printf '..%s..' "$foo".
>                   ^-- [SC2145](https://www.shellcheck.net/wiki/SC2145) (error): Argument mixes string and array. Use * or separate argument.
>  
> [Line 175:](javascript:setPosition(175, 12))
>     printf "$@\n"
>            ^-- [SC2059](https://www.shellcheck.net/wiki/SC2059) (info): Don't use variables in the printf format string. Use printf '..%s..' "$foo".
>             ^-- [SC2145](https://www.shellcheck.net/wiki/SC2145) (error): Argument mixes string and array. Use * or separate argument.
>  
> [Line 218:](javascript:setPosition(218, 5))
>     local PATH_NORMALIZED=$(printf '%s' "$PATH" | sed -e "s|:~|:$HOME|g" | sed -e "s|^~|$HOME|")
>     ^-- [SC3043](https://www.shellcheck.net/wiki/SC3043) (warning): In POSIX sh, 'local' is undefined.
>           ^-- [SC2155](https://www.shellcheck.net/wiki/SC2155) (warning): Declare and assign separately to avoid masking return values.
>                           ^-- [SC2046](https://www.shellcheck.net/wiki/SC2046) (warning): Quote this to prevent word splitting.
>  
> [Line 221:](javascript:setPosition(221, 5))
>     local QUERY_NORMALIZED=$(printf '%s' "$1" | sed -e "s|^~|$HOME|")
>     ^-- [SC3043](https://www.shellcheck.net/wiki/SC3043) (warning): In POSIX sh, 'local' is undefined.
>           ^-- [SC2155](https://www.shellcheck.net/wiki/SC2155) (warning): Declare and assign separately to avoid masking return values.
>                            ^-- [SC2046](https://www.shellcheck.net/wiki/SC2046) (warning): Quote this to prevent word splitting.
>  
> [Line 228:](javascript:setPosition(228, 125))
>     DATA='{ "api_key": "'$POSTHOG_WASP_PUBLIC_API_KEY'", "type": "capture", "event": "install-script:run", "distinct_id": "'$RANDOM`date +'%s%N'`'", "properties": { "os": "'`get_os_info`'" } }'
> >>                                                                                                                          ^-- [SC3028](https://www.shellcheck.net/wiki/SC3028) (warning): In POSIX sh, RANDOM is undefined.
> >>                                                                                                                                 ^-- [SC2006](https://www.shellcheck.net/wiki/SC2006) (style): Use $(...) notation instead of legacy backticks `...`.
> >>                                                                                                                                                                           ^-- [SC2006](https://www.shellcheck.net/wiki/SC2006) (style): Use $(...) notation instead of legacy backticks `...`.
> 
> Did you mean: ([apply this](javascript:applyFixIndex([13,14])), apply [all SC2006](javascript:applyFixCode(2006)))
>     DATA='{ "api_key": "'$POSTHOG_WASP_PUBLIC_API_KEY'", "type": "capture", "event": "install-script:run", "distinct_id": "'$RANDOM$(date +'%s%N')'", "properties": { "os": "'$(get_os_info)'" } }'
>  
> [Line 243:](javascript:setPosition(243, 36))
>         curl -LIs -o /dev/null -w %{url_effective} https://github.com/wasp-lang/wasp/releases/latest | awk -F/ '{print $NF}' | cut -c2-
>                                    ^-- [SC1083](https://www.shellcheck.net/wiki/SC1083) (warning): This { is literal. Check expression (missing ;/\n?) or quote it.
>                                                  ^-- [SC1083](https://www.shellcheck.net/wiki/SC1083) (warning): This } is literal. Check expression (missing ;/\n?) or quote it.
> 
> $
> ```
> 
> * Did we always install the data files, in addition to the executable, before? What do we need those for again? I thought that was packaged into the binary.
> * Do we want to have a nicer error message when they try to install a version that cannot be found?
> 
> <img alt="Screenshot 2022-11-01 104823" width="862" src="https://user-images.githubusercontent.com/523636/199262169-3d169014-fb31-442e-8bde-a8f446d9dbeb.png">
> 
> Overall, awesome job! I also left a couple of small ideas on tweaking the language, but looking great! rocket

Awesome @shayneczyzewski thanks the review, I completely forgot about shellcheck! But the way cool fact, in case you don't know: shellcheck is 100% written in Haskell!

So I run shellcheck and fixed all that I knew how to fix easily. There was couple of things that I didn't know how to fix, and concluded it is not worth spending time now on them if they work, so maybe we can handle those in the future.

I modified the code to print nice message on 404, assuming that version does not exist.

Data files -> yes, we always installed them next to binary. It would be nice if we could package them into binary itself, but that is a bit harder so we went with this since it was easier to set up. Maybe we can package them in the future, that would make the whole thing a bit simpler installation-wise.

Martinsos commented:
> Oh wow, awesome that shellcheck is Haskell! Seems like a useful tool. I didn't understand most of their suggestions, so glad you were able to fix a few! :D
> 
> Ok, thanks for confirming data files have not changed. 👍🏻
> 
> On the 404 front, I did notice one small oddity @Martinsos. It seems to show the nice message when curled, but not when executed locally. I think that is probably fine since most will do it the former, but do we want to support the latter as well? <img alt="Screenshot 2022-11-03 104527" width="854" src="https://user-images.githubusercontent.com/523636/199752652-aac4644b-a1a0-43a4-b0ff-1b17940546d1.png">
> 
> BTW, love the successful output on how to get a first Wasp app up, nice! star_struck

Uf ok this one took me a bit :D.
The thing is, it works for me  locally if I do `dash installer.sh` (that is how I test). I tried and it worked also for `bash installer.sh` and `sh installer.sh` so then I was very confused. But it does work as you said if I do `./installer.sh`. I realized at the end that the shebang is `#!/bin/sh -e`, and this `-e` told script to die if un untested (not used in if, && and similar) command has an error. So this caused it to stop early. I removed it now, I don't see the purpose for it, so this should be consistent now!

Martinsos commented:
Ok I think this should be it @shayneczyzewski ! I made it ready for review -> the only reason it was in draft was because I wanted to add docs, but I realized I need toa dd them in another place, so if you approve I will merge!

shayneczyzewski reviewed (approved):
LGTM, let's ship it! 🚀 Really nice work here, this will be super handy to have moving forward. Also, good find on the way of running the shell script impacts things... sorry for the rabbit hole on that one. :D

Martinsos commented:
Awesome merged it! Hah don't apologize for testing and fixing stuff :D.
